src/pip/_internal/models/selection_prefs.py

```diff
@@ -1,53 +1,32 @@
+from dataclasses import dataclass
 from typing import Optional
 
 from pip._internal.models.format_control import FormatControl
 
 
-# TODO: This needs Python 3.10's improved slots support for dataclasses
-# to be converted into a dataclass.
+@dataclass(frozen=True)
```

Comment:
I think it would be better to keep this as slots, since this is in the hot loop for the resolver and slots lookups tend to be ~10% quicker than non-slots lookups (based on my memory + rudimentary tests on Apple's arm64 + Python 3.12).

Reply:
:+1: I completely forgot about slotted attribute access being faster. For some reason, only the memory usage improvement stuck in my mind. I'll push an update when I am not procrastinating other (arguably more important) work!

Reply:
[`a69903a` (#12659)](https://github.com/pypa/pip/pull/12659/commits/a69903aced0f1810008f21e818e61ceb9fe49247)